src/util/mp_arith.h

```diff
@@ -48,4 +48,173 @@ const mp_integer string2integer(const std::string &, unsigned base=10);
+  {                                                                            \
+    return static_cast<const mp_integer &>(*this)                              \
+      op static_cast<const mp_integer &>(other);                               \
+  }                                                                            \
+  bool operator op(const mp_integer &other) const = delete
+
+template <typename Tag>
+class named_mp_integert
+{
+public:
+  explicit named_mp_integert(const mp_integer &generic) : value(generic)
+  {
+  }
+
+  explicit named_mp_integert(mp_integer &&generic) : value(std::move(generic))
+  {
+  }
+
+  const mp_integer &get() const
+  {
+    return value;
+  }
+
+  named_mp_integert &operator+=(const named_mp_integert &other)
+  {
+    value += other.value;
+    return *this;
+  }
+
+  named_mp_integert &operator-=(const named_mp_integert &other)
+  {
+    value -= other.value;
+    return *this;
+  }
+
+  named_mp_integert &operator%=(const named_mp_integert &other)
+  {
+    value %= other.value;
+    return *this;
+  }
+
+  bool operator<(const named_mp_integert &other) const
+  {
+    return value < other.value;
+  }
+
+  bool operator>(const named_mp_integert &other) const
+  {
+    return value > other.value;
+  }
+
+  bool operator<=(const named_mp_integert &other) const
+  {
+    return value <= other.value;
+  }
+
+  bool operator>=(const named_mp_integert &other) const
+  {
+    return value >= other.value;
+  }
+
+  bool operator==(const named_mp_integert &other) const
+  {
+    return value == other.value;
+  }
+
+  bool operator!=(const named_mp_integert &other) const
+  {
+    return value != other.value;
+  }
+
+private:
+  mp_integer value;
+};
+
+template <typename T>
+inline named_mp_integert<T>
+operator+(const named_mp_integert<T> &lhs, const named_mp_integert<T> &rhs)
+{
+  return named_mp_integert<T>{lhs} += rhs;
+}
+
+template <typename T>
+inline named_mp_integert<T>
+operator-(const named_mp_integert<T> &lhs, const named_mp_integert<T> &rhs)
+{
+  return named_mp_integert<T>{lhs} -= rhs;
+}
+
+template <typename T>
+inline named_mp_integert<T>
+operator*(const named_mp_integert<T> &lhs, const mp_integer &rhs)
+{
+  return named_mp_integert<T>{lhs.get() * rhs};
+}
+
+template <typename T>
+inline named_mp_integert<T>
+operator*(const mp_integer &lhs, const named_mp_integert<T> &rhs)
+{
+  return named_mp_integert<T>{lhs * rhs.get()};
+}
+
+template <typename T>
+inline named_mp_integert<T>
+operator%(const named_mp_integert<T> &lhs, const named_mp_integert<T> &rhs)
+{
+  return named_mp_integert<T>{lhs} %= rhs;
+}
+
+template <typename T>
+inline named_mp_integert<T>
+operator/(const named_mp_integert<T> &lhs, const mp_integer &rhs)
+{
+  return named_mp_integert<T>{lhs.get() / rhs};
+}
+
+template <typename T>
+inline mp_integer
+operator/(const named_mp_integert<T> &lhs, const named_mp_integert<T> &rhs)
+{
+  // division yields just a number
+  return lhs.get() / rhs.get();
```

Comment:
What would happen here if we divided by `0` by accident?

Reply:
`PRECONDITION`s added (did that here as well as for several other new functions).